File: apps/backend/app/api/budget.py
"""GET /api/research/budget — the read-only certification-budget accounting panel (goal-mcp-loop
iter-32, J-17 / backlog B-903).

Serves `app.engine.budget_accounting.build_budget_payload` verbatim (re-format only — no recompute):
total canonical trials to date, the current canonical `required_p` bar, the Thresholdout budget
remaining, and the staging LORD++ next-trial level — each with a per-trial spend-over-time series, all
re-read from the SAME `ledger` / `online_fdr` / `referee` seams `app.mcp.tools.verify_edge` uses.

No DB/session is needed (both ledgers are append-only state files, not the snapshot DB). Ledger paths
are config/env-driven via the existing resolvers (anti-goal: No magic numbers — no path literal here).
A missing/empty ledger (either or both) returns 200 with the honest empty-ledger accounting the
formulas naturally produce, never a 500 (anti-goal: resilience to data-shape change).

READ-ONLY, always: no proven-language — trial counts and alpha figures are descriptive accounting,
never a "Proven"/"Not yet proven" signal. That continues to flow solely from `app.engine.evidence` /
`GET /api/evidence`, untouched here.
"""
from __future__ import annotations

from fastapi import APIRouter

from app.engine.budget_accounting import build_budget_payload

router = APIRouter(tags=["budget"])


@router.get("/research/budget")
def get_budget() -> dict:
    """The certification-budget accounting payload, verbatim: `{"canonical": {...}, "staging": {...}}`.
    READ-ONLY — recomputes nothing beyond the two forward next-trial bars (via the SAME seams
    `verify_edge` uses). A missing/empty ledger (either or both) ⇒ the honest empty-ledger accounting
    (200, never 500)."""
    return build_budget_payload()

File: apps/frontend/lib/budget.ts
/**
 * Certification-budget accounting types (goal-mcp-loop iter-32, J-17 / backlog B-903).
 *
 * Mirrors `lib/graveyard.ts`'s types-only pattern for the SEPARATE `GET /api/research/budget` payload —
 * how much statistical-credibility budget has already been spent, re-read VERBATIM (or re-derived via
 * the SAME referee/ledger seams the certifier uses; re-format only — nothing recomputed here).
 *
 * This module carries NO proven-language: trial counts and alpha figures are descriptive accounting,
 * never a "Proven"/"Not yet proven" signal. The ONLY source of "Proven" stays the certified-claims
 * ledger via `lib/evidence.ts` / `GET /api/evidence`; this file never touches that path.
 */

/** One point on a ledger's per-trial spend-over-time series, read VERBATIM from that trial's OWN
 *  recorded verdict (never recomputed). `required_p` is the significance bar (Bonferroni or LORD++)
 *  that trial was actually judged at; `deflation_divisor` / `alpha_charged` ride along on the canonical
 *  series only (staging's `deflation_divisor` mirrors the trial ordinal under LORD++, not a meaningful
 *  divisor, so it is omitted there). */
export interface BudgetSpendPoint {
  trial: number;
  register_date: string | null;
  status: string | null;
  required_p: number | null;
  deflation_divisor?: number | null;
  alpha_charged?: number | null;
}

/** The canonical (strict-Bonferroni) accounting: trials to date (a DISPLAY value, distinct from the
 *  forward-looking `n_trials_next`), the forward next-trial bar, and the Thresholdout budget remaining. */
export interface CanonicalBudget {
  n_trials_to_date: number;
  n_trials_next: number;
  alpha_per_test: number;
  required_p: number;
  alpha_budget_total: number;
  alpha_spent: number;
  alpha_budget_remaining: number;
  spend_over_time: BudgetSpendPoint[];
}

/** The staging (LORD++) accounting: trials to date, and the forward next-trial significance level (the
 *  "alpha-wealth" figure) — the internal exploration economy, never served on the canonical /evidence bar. */
export interface StagingBudget {
  n_trials_to_date: number;
  n_trials_next: number;
  next_level: number;
  spend_over_time: BudgetSpendPoint[];
}

/** The `GET /api/research/budget` payload: the canonical + staging accounting, each self-contained. */
export interface BudgetResponse {
  canonical: CanonicalBudget;
  staging: StagingBudget;
}
